### PR TITLE
First commit for a pytorch-cpu package

### PR DESCRIPTION
**Pull request by hmaarrfk**
Pytorch needs the new compilers to work.
As is, no package will get created when this feedstock gets merged, but 

my personal feedstocks
https://github.com/ramonaoptics/pytorch-cpu-feedstock
https://dev.azure.com/ramonaoptics/feedstock-builds/_build/results?buildId=13

Show that once the compiler upgrade gets issue for this recipe, it will likely work.

- Windows: Requires Visual Studio 2017 image to and compilers from defaults: https://github.com/conda-forge/staged-recipes/pull/7430
- OSX:  
   - [ ] Xcode greater or equal to 8: https://github.com/conda-forge/staged-recipes/pull/7428 This is already the default when you rerender modern recipes.
   - [ ] Needs extra time to build. The build takes about 1:30 on Azure
- Linux: Needs the new compilers only, not toolchain.

For this reason, it really makes no sense to continue work on this recipe in staged-recipes

Also, I know this versions a lot of software, but short of redoing their whole build system, there isn't much I can do :/

In fact, in their release, they actually hash and patch all the libraries so that only the libraries that were compiled during the build get used by the build.

### CODE REVIEW

**Comment by sodre** [comment]
@hmaarrfk, I was working on that package tonight as well. Let me know if you need help. This is the version that I update  #3780, and here it is in my own branch https://github.com/sodre/staged-recipes/tree/ds/pytorch

**Comment by hmaarrfk** [comment]
I glanced at #3780, but figured that since it was open about 1 year ago, it would be stale since pytorch moves so fast.

Well, seems like VS2015 is no longer supported but there was a patch that claims to support.
I can't apply on my local machine so :/
https://github.com/pytorch/pytorch/pull/6854

But finally, maybe somebody know why linux is failing with:
```
    /tmp/pip-req-build-66DJyf/c10/test/impl/InlineStreamGuard_test.cpp: In function ‘c10::DeviceIndex& _ZTWN3c104impl13FakeGuardImplIXT_EE15current_device_E()’:
    /tmp/pip-req-build-66DJyf/c10/test/impl/InlineStreamGuard_test.cpp:174:1: internal compiler error: in var_defined_without_dynamic_init, at cp/decl2.c:2811
     }
```

**Comment by hmaarrfk** [comment]
@sodre I can't seem to create my own repo. Typically I can register things with:
https://github.com/ramonaoptics/pytorch-cpu-feedstock

but for some reason, circle is giving me troubles.

Basically, we need to upgrade gcc. 4.8 or whatever is too old.
Visual studio can **likely** get upgraded, but it seems that Anaconda has had lots of troubles with it. I figured we would start with linux anyway, as you did.

**Comment by hmaarrfk** [comment]
On travis, the build times out, on azure, I get the following error for osx:

```
2018-12-25T19:03:23.3573270Z     [ 85%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/anomaly_mode.cpp.o
2018-12-25T19:03:23.3573760Z     [ 85%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/engine.cpp.o
2018-12-25T19:03:23.3574230Z     [ 85%] Building CXX object caffe2/torch/lib/THD/CMakeFiles/THD.dir/base/init_methods/InitMethod.cpp.o
2018-12-25T19:03:23.3574710Z     [ 85%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/ps_roi_pool_op.cc.o
2018-12-25T19:03:23.3575890Z     In file included from /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/engine.cpp:1:
2018-12-25T19:03:23.3576980Z     In file included from /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/engine.h:7:
2018-12-25T19:03:23.3578350Z     In file included from /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/function.h:6:
2018-12-25T19:03:56.1461190Z Command "/Users/vsts/agent/2.144.0/work/1/a/miniconda/conda-bld/pytorch-cpu_1545761271021/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin/python -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-req-build-ia38pfyy/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/tmp/pip-record-0istyfxa/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /private/tmp/pip-req-build-ia38pfyy/
2018-12-25T19:03:58.6721520Z     /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/profiler.h:46:17: error: use of undeclared identifier 'CLOCK_MONOTONIC'
2018-12-25T19:03:58.6721830Z       clock_gettime(CLOCK_MONOTONIC, &t);
2018-12-25T19:03:58.6721920Z                     ^
2018-12-25T19:03:58.6722000Z     [ 85%] Building CXX object caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/function.cpp.o
2018-12-25T19:03:58.6722060Z     1 error generated.
2018-12-25T19:03:58.6722130Z     make[2]: *** [caffe2/torch/CMakeFiles/torch.dir/csrc/autograd/engine.cpp.o] Error 1
2018-12-25T19:03:58.6722200Z     make[2]: *** Waiting for unfinished jobs....
2018-12-25T19:03:58.6722270Z     [ 85%] Building CXX object modules/detectron/CMakeFiles/caffe2_detectron_ops.dir/roi_pool_f_op.cc.o
2018-12-25T19:03:58.6722860Z     In file included from /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/function.cpp:1:
2018-12-25T19:03:58.6723410Z     In file included from /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/function.h:6:
2018-12-25T19:03:58.6724010Z     /tmp/pip-req-build-ia38pfyy/torch/csrc/autograd/profiler.h:46:17: error: use of undeclared identifier 'CLOCK_MONOTONIC'
2018-12-25T19:03:58.6724100Z       clock_gettime(CLOCK_MONOTONIC, &t);
2018-12-25T19:03:58.6724170Z                     ^
2018-12-25T19:03:58.6724240Z     [ 85%] Building CXX object caffe2/torch/lib/THD/CMakeFiles/THD.dir/base/init_methods/InitMethodEnv.cpp.o
2018-12-25T19:03:58.6724300Z     1 error generated.
```

**Comment by jakirkham** [comment]
Yeah, `clock_gettime` was added in macOS 10.12, but did not exist previously. Currently we are building against macOS 10.9. This is probably worth raising an issue upstream. It's easy enough to workaround in their codebase if they are agreeable. There's some good discussion about this problem generally in issue ( https://github.com/zeromq/libzmq/issues/2175 ) and this [SO question]( https://stackoverflow.com/q/5167269 ).

File: .travis.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ jobs:
   include:
     - if: (fork IN (true)) OR (type IN (pull_request))
       os: osx
-      osx_image: xcode6.4
+      osx_image: xcode9.4
```
Comment (jakirkham) [inline .travis.yml]: This can be submitted in an independent PR, which we should merge promptly.

xref: https://github.com/conda-forge/conda-forge.github.io/issues/641

File: .travis_scripts/build_all [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
+set -ex
 
 # Fast finish the PR.
 echo ""
 (curl https://raw.githubusercontent.com/conda-forge/conda-forge-ci-setup-feedstock/master/recipe/ff_ci_pr_build.py | \
     python - -v --ci "travis" "${TRAVIS_REPO_SLUG}" "${TRAVIS_BUILD_NUMBER}" "${TRAVIS_PULL_REQUEST}") || exit 1
 
 # Set the numpy variable. This isn't used, but conda-build complains if we haven't set it already.
-export CONDA_NPY=19
+export CONDA_NPY=111
```
Comment (jakirkham) [inline .travis_scripts/build_all]: I think this hack is no longer needed. It predates conda-build 3. Would submit a separate PR to just remove it.

File: .travis_scripts/build_all [hunk 3]
```diff
@@ -29,7 +30,7 @@ echo ""
 echo "Configuring conda."
 conda config --add channels conda-forge
 conda config --set show_channel_urls true
-conda install --yes --quiet conda-forge-ci-setup=1.* conda-forge-pinning networkx conda-build=3.16
+conda install --yes --quiet conda-forge-ci-setup=2.* conda-forge-pinning networkx conda-build
```
Comment (jakirkham) [inline .travis_scripts/build_all]: Relaxing the `conda-build` version should go in a separate PR. I think all issues that caused this pinning were resolved, but it would be good to check with core in that PR in case others know of issues.

Comment (jakirkham) [inline .travis_scripts/build_all]: Bumping the `conda-forge-ci-setup` version constraint seems reasonable. Would submit a PR and request core's feedback to see if there is something else we are missing or not.

File: recipes/pytorch-cpu/conda-forge.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+compiler_stack: comp7
```
Comment (jakirkham) [inline recipes/pytorch-cpu/conda-forge.yml]: I think we build against both compilers in `staged-recipes` now. So we just need to skip the `toolchain` build. Adding this shouldn't be necessary. @isuruf, does this sound correct to you?

**Comment by hmaarrfk** [comment]
@jakirkham Thank you so much for your review. 

I think we should probably just merge this once I remove any changes not within the recipe since these builds will never pass here due to time constraints. 
Ill try to link to build results from my own feedstocks tomorrow

It seems there is finally interest in building with older compilers upstream 

https://github.com/pytorch/pytorch/issues/15294

Even though an earlier attempt to get it to compile with MSVC 2015 was pretty much shutdown outright. 

Do you know what version of Xcode defaults uses?

File: .travis_scripts/build_all (hunk 3 above)
Comment (jakirkham) [reply .travis_scripts/build_all]: The last issue I remember was Go related. If you think we are good to unpin it, am happy to move forward with that PR.

File: .travis_scripts/build_all (hunk 2 above)
Comment (hmaarrfk) [reply .travis_scripts/build_all]: Thanks: https://github.com/conda-forge/staged-recipes/pull/7427

**Comment by hmaarrfk** [comment]
@jakirkham  I think it only builds with the toolchain compilers.

Anyway, here is my own build that shows results for meta.yaml

https://github.com/ramonaoptics/pytorch-cpu-feedstock
Azure build
https://dev.azure.com/ramonaoptics/feedstock-builds/_build?definitionId=4

I think the rest of the work can be done outside of staged-recipes due to build times.

File: .travis_scripts/build_all (hunk 3 above)
Comment (hmaarrfk) [reply .travis_scripts/build_all]: Thanks: https://github.com/conda-forge/staged-recipes/pull/7429

**Comment by hmaarrfk** [comment]
It seems that my MacOSX patch has gotten the compiler to build the package:
https://dev.azure.com/ramonaoptics/feedstock-builds/_build/results?buildId=13

**Comment by sodre** [comment]
> @jakirkham I think it only builds with the toolchain compilers.
> 
In staged recipes we only use the toolchain compiler. After the feedstock is merged the bot will issue a PR to bump you up to GCC7.

File: recipes/pytorch-cpu/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+# The build needs to set many environment variables
+# This is an easy way to do it cross platform
+{% set export = "export" %}  # [not win]
+{% set export = "set" %}     # [win]
+
+build:
+  number: 1000
+  # Needs modern GCC, the ones in conda-forge are too old
+  {% if 'toolchain' in compiler('cxx') %}skip: True{% endif %}  # Only build with modern compilers.
```
Comment (sodre) [inline recipes/pytorch-cpu/meta.yaml]: I think this disabled the build of the package altogether in staged recipes.

Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: it will simply fail. It needs modern compilers, as well as few other PRs fixing old OSX stuff.

File: recipes/pytorch-cpu/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  patches:
+    - mac_osx_clock_getttime.patch  # [osx]
+
+# The build needs to set many environment variables
+# This is an easy way to do it cross platform
+{% set export = "export" %}  # [not win]
+{% set export = "set" %}     # [win]
+
+build:
+  number: 1000
```
Comment (sodre) [inline recipes/pytorch-cpu/meta.yaml]: This should go back to 0.

Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: Ok, but this kind of stuff should really just be worked on in the feedstock. No builds are being set, and we can spend for ever making things "tidy".

File: recipes/pytorch-cpu/meta.yaml (hunk 5 above)
Comment (sodre) [reply recipes/pytorch-cpu/meta.yaml]: <no judgement>Are you asking to import an empty feedstock so all build issues, including linux, osx, and win can be fixed outside staged-recipes? </no judgement>

If that is the case it might make it easier to simply say 
```yaml
build:
  number: 0
  skip: True 
```

File: recipes/pytorch-cpu/meta.yaml (hunk 6 above)
Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: I mean, no actual package is going to get built once this feedstock gets created, so it is pretty safe to "accept now" and fixup later.

Especially since we might have to request more time on Travis anyway.

File: recipes/pytorch-cpu/meta.yaml (hunk 5 above)
Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: I'm not, I am asking to import a recipe that I tested mostly with conda-smithy tools that I rendered here: https://github.com/ramonaoptics/pytorch-cpu-feedstock

https://dev.azure.com/ramonaoptics/feedstock-builds/_build/results?buildId=13

There is a pretty high probability that it will work once the bot issues the compiler upgrade.

File: recipes/pytorch-cpu/meta.yaml (hunk 6 above)
Comment (sodre) [reply recipes/pytorch-cpu/meta.yaml]: I've never merged a recipe like this one before so I rather defer the merge to @conda-forge/core.

File: recipes/pytorch-cpu/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,83 @@
+{% set version = "1.0.0" %}
+
+package:
+  name: pytorch-cpu
```
Comment (sodre) [inline recipes/pytorch-cpu/meta.yaml]: Sorry I did not catch this the first time...

Shouldn't the base recipe be called pytorch, and then have an output called pytorch-cpu?

Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: Good point. I thought about that possibility, but in my mind: 

1. split packages have been a pain to get approved.
2. Build variants in conda-forge don't do well with the linter. https://github.com/conda-forge/conda-smithy/issues/930 https://github.com/conda-forge/conda-smithy/issues/891
2. Conda forge doesn't compile with or link to any GPU stuff. Therefore, I really don't people thinking they might be installing `pytorch-gpu` accidentally.
3. Keeping this name makes it really easy to later upgrade build variants where a gpu and cpu version might coexist in conda-forge.

We can eventually deprecate this feedstock (pytorch-cpu) if pytorch-gpu becomes a thing for conda-forge.

File: recipes/pytorch-cpu/meta.yaml (hunk 5 above)
Comment (isuruf) [reply recipes/pytorch-cpu/meta.yaml]: To use new compilers on staged-recipes, add
```yaml
channel_sources:
- conda-forge/label/gcc7,defaults   # [unix]
- conda-forge,defaults              # [win]
channel_targets:
- conda-forge gcc7                  # [unix]
- conda-forge main                  # [win]
cxx_compiler:
- gxx      # [linux]
- clangxx  # [osx]
```

Comment (sodre) [reply recipes/pytorch-cpu/meta.yaml]: Where should we add these?

Comment (isuruf) [reply recipes/pytorch-cpu/meta.yaml]: to `recipes/pytorch-cpu/conda_build_config.yaml`

Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: thanks for chiming in @isuruf!

Unfortunately, this doesn't solve any of the other outstanding issues here:
- The need for VS2017, which currently isn't being used in staged-recipes. This is a change in appveyor that needs to be done.
- The need to increase the timeout in travis.

If I do this, the only package that will be built is the linux one, which really isn't the "attractiveness" of conda-forge.

Please merge this package as is, so the rest of the work can be done in the feedstock.

Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: ok will do.

File: recipes/pytorch-cpu/meta.yaml (hunk 6 above)
Comment (hmaarrfk) [reply recipes/pytorch-cpu/meta.yaml]: I had to put it back to 1000 since we are now building with the gcc compilers on linux.

**Comment by hmaarrfk** [comment]
@sodre well after 3 hours, it seems like a successful build.

**Comment by sodre** [comment]
🤞

**Comment by jakirkham** [comment]
>> I think it only builds with the toolchain compilers.
>
> In staged recipes we only use the toolchain compiler. After the feedstock is merged the bot will issue a PR to bump you up to GCC7.

Sorry, thought we had configured `staged-recipes` to do both.
